dig.go

```diff
@@ -314,6 +314,47 @@ type Container struct {
 
 	// invokerFn calls a function with arguments provided to Provide or Invoke.
 	invokerFn invokerFn
+
+	gh *graphHolder
+}
+
+// graphHolder represents the dependency graph for a Container. Specifically,
+// it saves constructorNodes and paramGroupedSlices (value groups) as graphNodes
+// and implements the Graph interface defined in internal/graph to run graph
+// algorithms on it. It has a 1-to-1 correspondence with a Container whose graph
+// it represents.
+type graphHolder struct {
+	// all the nodes defined in the graph.
+	allNodes []*graphNode
+
+	// Maps each graphNode to its order.
+	orders map[key]int
+
+	// Container whose graph this holder contains.
+	c *Container
+}
+
+func (gh *graphHolder) Order() int {
+	return len(gh.allNodes)
+}
+
+func (gh *graphHolder) EdgesFrom(u int) []int {
+	n := gh.allNodes[u]
+
+	var orders []int
+
+	switch w := n.Wrapped.(type) {
+	case *constructorNode:
+		for _, param := range w.paramList.Params {
+			orders = append(orders, getParamOrder(gh, param)...)
+		}
+	case *paramGroupedSlice:
```

Comment:
It feels like breaking a boundary here to use paramGroupedSlice as the graph node
because a value group technically has two ends: param and result.

Would it be reasonable to declare a separate `valueGroupNode` type for this?

```go
type valueGroupNode struct {
  Group string
  Type reflect.Type
}
```